spread.yaml

```diff
@@ -1,40 +1,124 @@
+    systems:
+      - ubuntu-20.04-64:
+          image: ubuntu-2004-64-virt-enabled
+          storage: 20G
+
+  qemu-nested:
+    type: qemu
+    # TODO:UC22: uc22 needs 2G or grub will not loopback the kernel snap
+    #            we use nested VM's, so use 4G for the host until we can boot
+    #            uc22 with less RAM
+    memory: 4G
+    systems:
+      - ubuntu-20.04-64:
+          username: ubuntu
+          password: ubuntu
+
+  lxd-nested:
+    type: lxd
+    container-profiles: core22
+    systems:
+      - ubuntu-20.04:
+          image: ucspread
+
+path: /home/core22
+
 exclude:
-    - "*.raw"
-    - "*.snap"
-    - Makefile
-    - README.md
-    - hooks
-    - parts
-    - prime
-    - snap
-    - spread.yaml
-    - stage
-kill-timeout: 20m
+  - .git
+  - "*.o"
+  - "*.a"
+  - "*.snap"
+  - "*.img"
+
 prepare: |
-    true
-restore: |
-    true
+  # NOTE: This part of the code needs to be in spread.yaml as it runs before
+  # the rest of the source code (including the tests/lib directory) is
+  # around. The purpose of this code is to fix some connectivity issues setup 
+  # spread functions to be more usable outside of the explicit task.yaml 
+  # sections.
+
+  # apt update is hanging on security.ubuntu.com with IPv6, prefer IPv4 over IPv6
+  cat <<EOF > gai.conf
+  precedence  ::1/128       50
+  precedence  ::/0          40
+  precedence  2002::/16     30
+  precedence ::/96          20
+  precedence ::ffff:0:0/96 100
+  EOF
+  if ! mv gai.conf /etc/gai.conf; then
+      echo "/etc/gai.conf is not writable, ubuntu-core system? apt update won't be affected in that case"
+      rm -f gai.conf
+  fi
+
+  # Take the MATCH and REBOOT functions from spread and allow our shell
+  # scripts to use them as shell commands. The replacements are real
+  # executables in tests/lib/bin (which is on PATH) but they source
+  # spread-funcs.sh written here, base on the definitions provided by SPREAD.
+  # This ensures that 1) spread functions define the code 2) both MATCH and
+  # REBOOT are executables and not functions, and can be called from any
+  # context.
+  type MATCH | tail -n +2 > "$TESTSLIB"/spread-funcs.sh
+  unset MATCH
+  type REBOOT | tail -n +2 >> "$TESTSLIB"/spread-funcs.sh
+  unset REBOOT
+
+  # prepare common uc image setup by repacking snaps, etc
+  "$TESTSLIB"/prepare-uc.sh
+restore-each: |
+  # delete the nested VM image after each task finishes so we don't use too much
+  # disk space
+  . "$EXTTESTSLIB/nested.sh"
+  cleanup_nested_core_vm
+debug-each: |
+  . "$EXTTESTSLIB/nested.sh"
+  print_nested_status
+
+warn-timeout: 40m
+kill-timeout: 50m
+
 suites:
-    tests/main/:
-        summary: Tests for core20 snap.
+  tests/spread/:
+    summary: Tests for core-base snap.
+    environment:
+      # since we are only testing uc20 here, disable kvm unless the env
+      # specifically knows kvm works and sets this to true
+      # kvm specifically does not work on GCE with focal L1 VM and focal (UC20)
```

Comment:
Actually the comment here is referring to the outer VM which is indeed ubuntu 20.

About your previous comment about the lib/external, this is actually an imported subtree, subtrees do not appear as external in git, but rather as an integrated part of the repository which you then update via a command  `git subtree pull <repo> <path>`.

Reply:
Right, but it looks like `lib/external` is added in a commit mixed with other changes, while in theory `git subtree add --squash --prefix ...` should create a separate commit. I think it is important to keep that commit because afaik `git subtree` greps through commit log to know subtree history.

Reply:
Ah really good point let me fix that